File: music-commons-core/src/test/java/net/transgressoft/commons/music/StandardMusicLibraryTest.java
package net.transgressoft.commons.music;

import net.transgressoft.commons.event.QueryEventDispatcher;
import net.transgressoft.commons.music.audio.AudioItem;
import net.transgressoft.commons.music.audio.AudioItemRepository;
import net.transgressoft.commons.music.playlist.AudioPlaylist;
import net.transgressoft.commons.music.playlist.AudioPlaylistDirectory;
import net.transgressoft.commons.music.playlist.AudioPlaylistRepository;
import net.transgressoft.commons.music.waveform.AudioWaveform;
import net.transgressoft.commons.music.waveform.AudioWaveformRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Collections;
import java.util.Optional;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class StandardMusicLibraryTest {

    MusicLibrary<AudioItem, AudioPlaylist<AudioItem>, AudioPlaylistDirectory<AudioItem>, AudioWaveform> musicLibrary;

    @Mock
    AudioItemRepository<AudioItem> audioItemRepository;
    @Mock
    AudioPlaylistRepository<AudioItem, AudioPlaylist<AudioItem>, AudioPlaylistDirectory<AudioItem>> audioPlaylistRepository;
    @Mock
    AudioWaveformRepository<AudioWaveform> audioWaveformRepository;
    @Spy
    QueryEventDispatcher<AudioItem> audioItemEventDispatcher;

    @Test
    @DisplayName("Music api test")
    void musicApiTest() throws Exception {
        when(audioItemRepository.iterator()).thenReturn(Collections.emptyIterator());

        musicLibrary = StandardMusicLibrary.INSTANCE;

        verify(audioItemRepository).iterator();
        verify(audioItemEventDispatcher).subscribe(musicLibrary.getAudioItemSubscriber());

        musicLibrary.deleteAudioItems(Collections.singleton(mock(AudioItem.class)));
        verify(audioItemRepository).removeAll(any());
        verify(audioPlaylistRepository).removeAudioItems(any());
        verify(audioWaveformRepository).findById(anyInt());
        verifyNoMoreInteractions(audioItemEventDispatcher);

        musicLibrary.addAudioItemsToPlaylist(Collections.emptyList(), mock(AudioPlaylist.class));
        verify(audioPlaylistRepository).addAudioItemsToPlaylist(any(), any());
        verifyNoMoreInteractions(audioItemEventDispatcher);

        musicLibrary.removeAudioItemsFromPlaylist(Collections.emptyList(), mock(AudioPlaylist.class));
        verify(audioPlaylistRepository).removeAudioItemsFromPlaylist(any(), any());
        verifyNoMoreInteractions(audioItemEventDispatcher);

        musicLibrary.movePlaylist(mock(AudioPlaylist.class), mock(AudioPlaylistDirectory.class));
        verify(audioPlaylistRepository).movePlaylist(any(), any());

        when(audioWaveformRepository.findById(anyInt())).thenReturn(Optional.of(mock(AudioWaveform.class)));
        var result = musicLibrary.getOrCreateWaveformAsync(mock(AudioItem.class), (short) 500, (short) 150);
        assertThat(result.get()).isNotNull();

        verifyNoMoreInteractions(audioItemRepository, audioPlaylistRepository, audioWaveformRepository, audioItemEventDispatcher);
    }
}

File: music-commons-core/src/test/kotlin/net/transgressoft/commons/music/StandardMusicLibraryTest.kt
package net.transgressoft.commons.music

import com.google.common.truth.Truth.assertThat
import net.transgressoft.commons.event.QueryEventDispatcher
import net.transgressoft.commons.music.audio.AudioItem
import net.transgressoft.commons.music.audio.AudioItemRepository
import net.transgressoft.commons.music.playlist.AudioPlaylist
import net.transgressoft.commons.music.playlist.AudioPlaylistDirectory
import net.transgressoft.commons.music.playlist.AudioPlaylistRepository
import net.transgressoft.commons.music.waveform.AudioWaveform
import net.transgressoft.commons.music.waveform.AudioWaveformRepository
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyShort
import org.mockito.kotlin.*
import java.util.*

internal class StandardMusicLibraryTest {

    val audioItemRepository: AudioItemRepository<AudioItem> = mock {
        on { iterator() } doReturn Collections.emptyIterator()
    }

    val audioWaveformRepository = mock<AudioWaveformRepository<AudioWaveform>> {
        on { findById(anyInt()) } doReturn Optional.of(mock {})
        on { findById(eq(9)) } doReturn Optional.empty()
        on { create(any<AudioItem>(), anyShort(), anyShort()) } doReturn mock {}
    }

    val audioPlaylistRepository =
        mock<AudioPlaylistRepository<AudioItem, AudioPlaylist<AudioItem>, AudioPlaylistDirectory<AudioItem, AudioPlaylist<AudioItem>>>> {}
    val audioItemEventDispatcher = spy<QueryEventDispatcher<AudioItem>> {}

    val musicLibrary = StandardMusicLibrary(audioItemRepository, audioPlaylistRepository, audioWaveformRepository, audioItemEventDispatcher)

    @Test
    fun `Music api test`() {
        verify(audioItemRepository).iterator()
        verify(audioItemEventDispatcher).subscribe(musicLibrary.audioItemSubscriber)

        musicLibrary.deleteAudioItems(setOf(mock {}))
        verify(audioItemRepository).removeAll(any())
        verify(audioPlaylistRepository).removeAudioItems(any())
        verify(audioWaveformRepository).findById(any())
        verify(audioWaveformRepository).remove(any())
        verifyNoMoreInteractions(audioItemEventDispatcher)

        musicLibrary.addAudioItemsToPlaylist(emptyList(), mock {})
        verify(audioPlaylistRepository).addAudioItemsToPlaylist(any(), any())
        verifyNoMoreInteractions(audioItemEventDispatcher)

        musicLibrary.removeAudioItemsFromPlaylist(emptyList(), mock {})
        verify(audioPlaylistRepository).removeAudioItemsFromPlaylist(any(), any())
        verifyNoMoreInteractions(audioItemEventDispatcher)

        musicLibrary.movePlaylist(mock {}, mock {})
        verify(audioPlaylistRepository).movePlaylist(any(), any())

        val result = musicLibrary.getOrCreateWaveformAsync(mock { on { id } doReturn 9 }, 500.toShort(), 150.toShort())
        assertThat(result.get()).isNotNull()
        verify(audioWaveformRepository).findById(eq(9))
        verify(audioWaveformRepository).create(any(), eq(500), eq(150))

        verifyNoMoreInteractions(audioItemRepository, audioPlaylistRepository, audioWaveformRepository, audioItemEventDispatcher)
    }
}